### ISSUE TITLE
[no-misused-promises] non-thenable thenable union react component attribute does not accept thenable function

### ISSUE THREAD

**Issue by ThiloAschebrock** [typescript-eslint/typescript-eslint#4650]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": ["error"],
  }
}
```

```TS
interface Props {
  onEvent: (() => void) | (() => Promise<void>);
}

const Component: React.FC<Props> = () => null;

const App: React.FC = () => {
  const handleEvent = async () => {};
  return <Component onEvent={handleEvent} />;
};

export default App;
```

```JSON
{
  "compilerOptions": {
    "target": "ESNext",
    "useDefineForClassFields": true,
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "allowJs": false,
    "skipLibCheck": false,
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react-jsx"
  },
  "include": ["src"],
  "references": [{ "path": "./tsconfig.node.json" }]
}
```
**Expected Result**

No linting error since `onEvent` accepts both a thenable and a non-thenable function.

**Actual Result**

```
9:29  error  Promise-returning function provided to attribute where a void return was expected  @typescript-eslint/no-misused-promises`
```
**Additional Information**

This bug was introduced in `5.14.0`. `5.13.0` ist working fine.

Note that you can circumvent the bug by using 
```ts
interface Props {
  onEvent: () => void | Promise<void>;
}
```
instead.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.14.0` |
| `@typescript-eslint/parser`        | `5.14.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `14.18.3` |

**Comment by ThiloAschebrock** [comment]
To me it rather looks like #4620 introduced this bug.

**Comment by JoshuaKGoldberg** [comment]
Indeed - this is another edge case I hadn't thought of. 😞

**Comment by kachkaev** [comment]
What would be the right way forward? Just wonder if I should refactor my React callbacks or just temporarily disable the rule for now.

**Comment by bradzacher** [comment]
@esetnik - I think you've misunderstood the OP.

[Passing `() => Promise<T>` to `() => void` is a lint error, but passing it to `(() => void) | (() => Promise<T>)` is not.](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZAtnFAK5QQAmZOhtT7UVAC4wAbVQRCvQqgC6YEMABQSjhHwwAhoQhgAZiwT4kcWgjAA3WnA4B5BDCwAKbQHMRTgJRgAvAD5Law5PESsbAG5VdS0dfUNjU3Mwu0IABV5+CBdCdzAnL18A5O8AHzyC-zB0vmYIAB5kv08QwIiVNQ1tXXwzKBQ9BA9vSurMur7iBFc-SKVk+0cnAc9I+bSM2qWEFZVdvf2Dw-3gYDAAYh1XRLBXGFoAI00YMB4a9jBsPCgP2jAAd1ohAA1tpaIYOB8ABa6HBaLCuXjgsD3FiuKAqUAgJQKMAAJQgPGgEGQ3yQ0LAPQYsIgJjMYFoejAmnMmigWCMkN4dDY9LwhE0tIQ2MUSkQSEkek0+F0o1qdQAKgEAN7YsBqzGqtU4gCCSCQUuh33wTxgj3wQO+ekBUN0OigtBgLEFTIQHGA1p0ACt1M6GTaqht2AA6TVq8DoHDaTQMekIAwwPRwGAwThgeXk43Js1An6QAAe6iduj+0PMZJlgd0zDAdodFk4Ia1YbAEajMbMXp9qfT3RN2dzEAL+CL-1L-tl72rneMDdDOPQOiQLEICG+2oDb10VsI-sp1N9jJLcHwkMg9Z3mdNUpz1cHhfFHEbTZFTfLCAV+NYMCQAEZfGmABo00-FhvwAJn-BAIHPPwnDMeNE2TTgAH4PCcCwnhYCARHlYYAnlEDvz-MoJwgAAZOAgXqAjoFA38mjAMpwQgRMoIhMoEFAmAgI7CBvRnDhUPKHRWTMERmSwPDgNo8DGI3TIKKoj8ZKQMCGKY10WMQVMOK4lpSOUr9fzkmijLU2Ymw1JsdT1A1oCZCk+2vfRrTMRx-WnA9x0rJ8tXDSN+XbBBPO7DMnPNAchxHEtiW8zcwCnPiu0fOdw0XZdVzAddSJcndywpPh92uP18svbNfLDZQm2NJAT0MujIOgyRYN4-iHyE-IRPtQYXUkwppKMuTSMU6jCKQdSwGY1idLATjk30ysFRMsaZiUABfUVkAlKUK03ABhXokEIFhjGtFVLJALFrPAdcdD0SRiWlXN8teJBaE+CAKoUKqtREjg3KwF5eDej6RAMiTVtDKznzAPaRPFb5NFmiA-nk2ovv8tt83va11zK5y2FTN6EoQOATCeOAAC9dBeny00hat8Yi6tIygdgISQAEmRyFgGGJJAoCEVKHNrGB60crMCfZ3NRfF2n4r+MnTyRjDHV0a18rtBq-WZd7oR3V5MgA4XmQhJHPIlq8IsJjnfgt+XMn+JWHNeCwbFTLq6WtSRpAxn61Sg1GFVgu9hzewgPFl7C8lVrCcKGysRuDqTkiAzzI4gUSECEiSU6CPObAWzdg4srVob82H4fspGcrJAUEu+KOIUVskXW5-lAb9LXv2+GKWWTelGXlt21AhUj0WumtoDrCAeJ3ELm7HCS0cnRukoEjGWwC6NLEw+ztRZKRNE7xlx83hcaQytdkdR0i-dDE0lsHcVXTXzR-ocQHDCBOg-gQUR5BlAAbBWO0AcJFzGEqMAZA-oAzEKkEmYAqJYD9PKWQIhtR-E0GTTgCpRCpFkAENakNLKnHYNHeA9wgzQDAgABh-AAViDDg-k9wUxBkfALKGpxH6KnQnvQWYAACS4pWEpiWiRROlFqJ+CaGDRa8pgGl3VJdYWcNM4IwcrXSE9dEr2jFqmD068HyjlisvEq5JXbuzHpWRGsQm7CyMW1Wck9WyBV3mrNch8O6DxXtAc+6UVzX0Dn4++TZ+TSifnmF+HA34f3ct-X+-9AFiEIfwzx4D5HF0wdgh8eDOIMHuJIQhJCy5kIgBQuAVCaH0KYSwzQbDPqcInmUmsO1g7pKwoIkRkgGniMVBAuU8p8wxO+MNaRdRED3R3AAVQCMhMAMywA4VKSoq6MN1ECmrjfKeziISG1qLlf0VjR5TyzufbeMZPbmB7Gc7qxzKz-FZLs5KATL5BKyjsheoTUr+xecYJaPhkYwScNcnOCBJJZLGIqZRYBy7Nk2ZopGISm5A03G8pcHzbrTwMfsnyvzQxRy8FCuUjRYXwpxIi7ZKKcX1jxfFbcDzaAj1TKAi5WNQGfIOcGYWF9MWZXXNy4skJaCTi2iuJ4YA+hbLAAwAUJ57LD2sWizIYTfq0pkZ06OwzJGbiTgMkl7AS7rSiJ0WIPRVwoFIoaiAB1LXHVOoQWYJxzjEg4Jca4twHiSqFR8XA0AgA&rules=N4XyA&tsConfig=N4XyA).
Which is expected, as the latter signature correctly identifies that the receiver expects a promise.

**Comment by esetnik** [comment]
> @bradzacher please review again the OP example which was working in v5.13.0 and broke in v5.14.0+. What's really weird is that I pasted this same example into the playground page you linked and it is working without error. But when I test this same example in my code using v5.17.0 I get the same linter error as OP. The only difference is that i'm using typescript v4.6.3 and the playground uses typescript v4.6.2.

<img width="500" alt="image" src="https://user-images.githubusercontent.com/664434/161289042-a65eda11-f34c-4b9b-addf-ee330a37e5c9.png">


```
interface Props {
  onEvent: (() => void) | (() => Promise<void>);
}

const Component: React.FC<Props> = () => null;

const App: React.FC = () => {
  const handleEvent = async () => {};
  return <Component onEvent={handleEvent} />;
};

export default App;
```

Note that `onEvent` is declared as `(() => void) | (() => Promise<void>)` and yet it generates the following error when an async function is passed:

```
9:29  error  Promise-returning function provided to attribute where a void return was expected  @typescript-eslint/no-misused-promises`
```

**Comment by esetnik** [comment]
@bradzacher just wanted to let you know I tested in v5.19.0 and it's still generating the spurious error. I believe this issue needs to be reopened.

**Comment by JoshuaKGoldberg** [comment]
Fix in review here: #4841

Since it's a bit buried in the thread, bumping that you can disable the new check added in v4.6 with the settings described at https://typescript-eslint.io/rules/no-misused-promises/#checksvoidreturn:

```json
"rules": {
  "@typescript-eslint/no-misused-promises": ["error", {
    "checksVoidReturn": {
      "arguments": false
    }
  }]
}
```

**Comment by esetnik** [comment]
> @esetnik - I think you've misunderstood the OP.
> 
> [Passing `() => Promise<T>` to `() => void` is a lint error, but passing it to `(() => void) | (() => Promise<T>)` is not.](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZAtnFAK5QQAmZOhtT7UVAC4wAbVQRCvQqgC6YEMABQSjhHwwAhoQhgAZiwT4kcWgjAA3WnA4B5BDCwAKbQHMRTgJRgAvAD5Law5PESsbAG5VdS0dfUNjU3Mwu0IABV5+CBdCdzAnL18A5O8AHzyC-zB0vmYIAB5kv08QwIiVNQ1tXXwzKBQ9BA9vSurMur7iBFc-SKVk+0cnAc9I+bSM2qWEFZVdvf2Dw-3gYDAAYh1XRLBXGFoAI00YMB4a9jBsPCgP2jAAd1ohAA1tpaIYOB8ABa6HBaLCuXjgsD3FiuKAqUAgJQKMAAJQgPGgEGQ3yQ0LAPQYsIgJjMYFoejAmnMmigWCMkN4dDY9LwhE0tIQ2MUSkQSEkek0+F0o1qdQAKgEAN7YsBqzGqtU4gCCSCQUuh33wTxgj3wQO+ekBUN0OigtBgLEFTIQHGA1p0ACt1M6GTaqht2AA6TVq8DoHDaTQMekIAwwPRwGAwThgeXk43Js1An6QAAe6iduj+0PMZJlgd0zDAdodFk4Ia1YbAEajMbMXp9qfT3RN2dzEAL+CL-1L-tl72rneMDdDOPQOiQLEICG+2oDb10VsI-sp1N9jJLcHwkMg9Z3mdNUpz1cHhfFHEbTZFTfLCAV+NYMCQAEZfGmABo00-FhvwAJn-BAIHPPwnDMeNE2TTgAH4PCcCwnhYCARHlYYAnlEDvz-MoJwgAAZOAgXqAjoFA38mjAMpwQgRMoIhMoEFAmAgI7CBvRnDhUPKHRWTMERmSwPDgNo8DGI3TIKKoj8ZKQMCGKY10WMQVMOK4lpSOUr9fzkmijLU2Ymw1JsdT1A1oCZCk+2vfRrTMRx-WnA9x0rJ8tXDSN+XbBBPO7DMnPNAchxHEtiW8zcwCnPiu0fOdw0XZdVzAddSJcndywpPh92uP18svbNfLDZQm2NJAT0MujIOgyRYN4-iHyE-IRPtQYXUkwppKMuTSMU6jCKQdSwGY1idLATjk30ysFRMsaZiUABfUVkAlKUK03ABhXokEIFhjGtFVLJALFrPAdcdD0SRiWlXN8teJBaE+CAKoUKqtREjg3KwF5eDej6RAMiTVtDKznzAPaRPFb5NFmiA-nk2ovv8tt83va11zK5y2FTN6EoQOATCeOAAC9dBeny00hat8Yi6tIygdgISQAEmRyFgGGJJAoCEVKHNrGB60crMCfZ3NRfF2n4r+MnTyRjDHV0a18rtBq-WZd7oR3V5MgA4XmQhJHPIlq8IsJjnfgt+XMn+JWHNeCwbFTLq6WtSRpAxn61Sg1GFVgu9hzewgPFl7C8lVrCcKGysRuDqTkiAzzI4gUSECEiSU6CPObAWzdg4srVob82H4fspGcrJAUEu+KOIUVskXW5-lAb9LXv2+GKWWTelGXlt21AhUj0WumtoDrCAeJ3ELm7HCS0cnRukoEjGWwC6NLEw+ztRZKRNE7xlx83hcaQytdkdR0i-dDE0lsHcVXTXzR-ocQHDCBOg-gQUR5BlAAbBWO0AcJFzGEqMAZA-oAzEKkEmYAqJYD9PKWQIhtR-E0GTTgCpRCpFkAENakNLKnHYNHeA9wgzQDAgABh-AAViDDg-k9wUxBkfALKGpxH6KnQnvQWYAACS4pWEpiWiRROlFqJ+CaGDRa8pgGl3VJdYWcNM4IwcrXSE9dEr2jFqmD068HyjlisvEq5JXbuzHpWRGsQm7CyMW1Wck9WyBV3mrNch8O6DxXtAc+6UVzX0Dn4++TZ+TSifnmF+HA34f3ct-X+-9AFiEIfwzx4D5HF0wdgh8eDOIMHuJIQhJCy5kIgBQuAVCaH0KYSwzQbDPqcInmUmsO1g7pKwoIkRkgGniMVBAuU8p8wxO+MNaRdRED3R3AAVQCMhMAMywA4VKSoq6MN1ECmrjfKeziISG1qLlf0VjR5TyzufbeMZPbmB7Gc7qxzKz-FZLs5KATL5BKyjsheoTUr+xecYJaPhkYwScNcnOCBJJZLGIqZRYBy7Nk2ZopGISm5A03G8pcHzbrTwMfsnyvzQxRy8FCuUjRYXwpxIi7ZKKcX1jxfFbcDzaAj1TKAi5WNQGfIOcGYWF9MWZXXNy4skJaCTi2iuJ4YA+hbLAAwAUJ57LD2sWizIYTfq0pkZ06OwzJGbiTgMkl7AS7rSiJ0WIPRVwoFIoaiAB1LXHVOoQWYJxzjEg4Jca4twHiSqFR8XA0AgA&rules=N4XyA&tsConfig=N4XyA). Which is expected, as the latter signature correctly identifies that the receiver expects a promise.

@JoshuaKGoldberg I see your fix and that makes a lot of sense. Just curious, do you have any idea why I'm not able to demonstrate this error in the playground?

**Comment by esetnik** [comment]
> 



> Fix in review here: #4841
> 
> Since it's a bit buried in the thread, bumping that you can disable the new check added in v4.6 with the settings described at https://typescript-eslint.io/rules/no-misused-promises/#checksvoidreturn:
> 
> ```json
> "rules": {
>   "@typescript-eslint/no-misused-promises": ["error", {
>     "checksVoidReturn": {
>       "arguments": false
>     }
>   }]
> }
> ```

I had to use:
```json
"rules": {
  "@typescript-eslint/no-misused-promises": ["error", {
    "checksVoidReturn": {
      "arguments": false,
      "attributes": false
    }
  }]
}
```
to silence all the situations affected by #4650
